### ISSUE TITLE
Error in bounding box when running with mixed scalar and array arguments.

### ISSUE THREAD

**Issue by WilliamJamieson** [astropy/astropy#12319]
### Description

When evaluating models using mixes of array and scalar arguments with the `with_bounding_box=True` option one gets an unexpected error in the bounding box logic.

Note that this happens on the current `main` branch and on astropy `v4.3.1`. So I do not think its an issue stemming from merging #11930 or #11931.

### Expected behavior
The model to evaluate without issue.

### Actual behavior
A value error due to misshaped arrays.

### Steps to Reproduce
On the astropy `main` branch the following code:
```python
from astropy.modeling import models
from astropy.modeling.bounding_box import BoundingBox
model = models.Shift(1) & models.Shift(2) & models.Identity(1)
model.inputs = ('x', 'y', 'slit_id')
bbox = BoundingBox.validate(model, ((-0.5, 1047.5), (-0.5, 2047.5), (-np.inf, np.inf)), order='F')
model.bounding_box = bbox
x = np.array([1000, 1001])
y = np.array([2000, 2001])

print(model(x, y, 0))
print(model(x, y, 0, with_bounding_box=True))
```
Produces the following error:
```python
ValueError                                Traceback (most recent call last)
<ipython-input-1-a1894efc3d83> in <module>
      9
     10 print(model(x, y, 0))
---> 11 print(model(x, y, 0, with_bounding_box=True))

~/projects/astropy/astropy/modeling/core.py in __call__(self, *args, **kwargs)
   1052         #   for CompoundModel(s).
   1053         # TODO: eliminate this in CompoundBoundingBox
-> 1054         outputs = self._generic_evaluate(evaluate, inputs,
   1055                                          fill_value, with_bbox,
   1056                                          (not isinstance(self, CompoundModel)))

~/projects/astropy/astropy/modeling/core.py in _generic_evaluate(self, evaluate, _inputs, fill_value, with_bbox, with_units)
   1014             input_shape = self._validate_input_shapes(_inputs, self._argnames, self.model_set_axis)
   1015
-> 1016             outputs = bbox.evaluate(evaluate, input_shape, _inputs, fill_value, with_units)
   1017         else:
   1018             outputs = evaluate(_inputs)

~/projects/astropy/astropy/modeling/bounding_box.py in evaluate(self, evaluate, input_shape, inputs, fill_value, with_units)
    612
    613     def evaluate(self, evaluate: Callable, input_shape, inputs, fill_value, with_units: bool=True):
--> 614         outputs, valid_outputs_unit = self._evaluate(evaluate, inputs, input_shape,
    615                                                      fill_value, with_units)
    616         outputs = self._set_valid_outputs_unit(outputs, valid_outputs_unit)

~/projects/astropy/astropy/modeling/bounding_box.py in _evaluate(self, evaluate, inputs, input_shape, fill_value, with_units)
    596     def _evaluate(self, evaluate: Callable, inputs, input_shape,
    597                   fill_value, with_units: bool):
--> 598         valid_inputs, valid_index, all_out = self.prepare_inputs(input_shape, inputs)
    599
    600         if all_out:

~/projects/astropy/astropy/modeling/bounding_box.py in prepare_inputs(self, input_shape, inputs)
    462         """
    463
--> 464         valid_index, all_out = self._valid_index(input_shape, inputs)
    465
    466         valid_inputs = []

~/projects/astropy/astropy/modeling/bounding_box.py in _valid_index(self, input_shape, inputs)
    432             if all of the inputs are outside the bounding_box
    433         """
--> 434         outside_index, all_out = self._outside(input_shape, inputs)
    435
    436         valid_index = np.atleast_1d(np.logical_not(outside_index)).nonzero()

~/projects/astropy/astropy/modeling/bounding_box.py in _outside(self, input_shape, inputs)
    409             else:
    410                 outside_index |= outside
--> 411                 if outside_index:
    412                     all_out = True
    413

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

In astropy `v4.3.1` the following (equivalent) code:
```python
from astropy.modeling import models
model = models.Shift(1) & models.Shift(2) & models.Identity(1)
model.inputs = ('x', 'y', 'slit_id')
model.bounding_box = ((-np.inf, np.inf), (-0.5, 2047.5), (-0.5, 1047.5))
x = np.array([1000, 1001])
y = np.array([2000, 2001])

print(model(x, y, 0))
print(model(x, y, 0, with_bounding_box=True))
```
Produces the following error:
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-887426bf0114> in <module>
----> 1 model(x, y, 0, with_bounding_box=True)

~/projects/astropy/astropy/modeling/core.py in __call__(self, *args, **kw)
   2826             input_shape = _validate_input_shapes(args, (), self._n_models,
   2827                                                  self.model_set_axis, self.standard_broadcasting)
-> 2828             vinputs, valid_ind, allout = prepare_bounding_box_inputs(self, input_shape, args, bbox)
   2829             if not allout:
   2830                 valid_result = self._evaluate(*vinputs, **kw)

~/projects/astropy/astropy/modeling/core.py in prepare_bounding_box_inputs(self, input_shape, inputs, bbox)
   4183         else:
   4184             nan_ind |= outside
-> 4185             if nan_ind:
   4186                 allout = True
   4187     # get an array with indices of valid inputs

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

```

### System Details

For running the code on astropy `main` my system details are:
```
macOS-10.15.7-x86_64-i386-64bit
Python 3.9.7 (default, Sep 17 2021, 14:07:07)
[Clang 12.0.0 (clang-1200.0.32.29)]
Numpy 1.21.3
pyerfa 2.0.0
astropy 5.0.dev1111+g032a9b5ac
Scipy 1.7.1
Matplotlib 3.4.3
```

For running the code on astropy `v4.3.1` my system details are:
```
macOS-10.15.7-x86_64-i386-64bit
Python 3.9.7 (default, Sep 17 2021, 14:07:07)
[Clang 12.0.0 (clang-1200.0.32.29)]
Numpy 1.21.3
pyerfa 2.0.0
astropy 4.3.1
Scipy 1.7.1
Matplotlib 3.4.3
```

**Comment by WilliamJamieson** [comment]
This bug was fixed for astropy `main` as part of PR #11942. However, due to the refactoring that occurred leading up to #11942 in PRs #11930 and #11931. These fixes cannot be directly backported to astropy 4; however, I can easily put together a PR for astropy 4 which will provide the same bugfix.

**Comment by nden** [comment]
Fixed in 5.0.
